### PR TITLE
Add custom macro editor for KaTeX plugin

### PR DESCRIPTION
**Pull request by 8d1h**
...as requested in https://github.com/Jermolene/TiddlyWiki5/issues/4375, with a nice UI :)

katex

**Update.** Added a few tweaks: no need to refresh for new macros now, and macros with arguments can be added with `#` (`\dd#1#2` in the above example).

### CODE REVIEW

File: plugins/tiddlywiki/katex/wrapper.js [hunk 1]
```diff
@@ -16,6 +16,19 @@ var katex = require("$:/plugins/tiddlywiki/katex/katex.min.js"),
     chemParse = require("$:/plugins/tiddlywiki/katex/mhchem.min.js"),
 	Widget = require("$:/core/modules/widgets/widget.js").widget;
 
+katex.macros = {};
+katex.updateMacros = function() {
+	var tid, macro, cmd;
+	const regex = /#\d/g; // Remove the arguments like #1#2
+	for (const t of $tw.wiki.getTiddlersWithTag("$:/tags/katex/macro")) {
```
Comment (pmario) [inline plugins/tiddlywiki/katex/wrapper.js:23]: In TW core we use ES5 javascript only. So you need to use `var` instead of `const` and the `for of` loop was introduced in ES6. So you need to use the "old way"

```
var tiddlers = $tw.wiki.getTiddlersWithTag("$:/tags/katex/macro") ;
for (var i=0; i < tiddlers.length; i++) {
//your code changed a bit
}
```

**Comment by pmario** [comment]
@8d1h did you sign the TW CLA already? https://github.com/Jermolene/TiddlyWiki5/blob/master/licenses/cla-individual.md If not, you should do so in a different PR.

File: plugins/tiddlywiki/katex/wrapper.js (hunk 1 above)
Comment (8d1h) [reply plugins/tiddlywiki/katex/wrapper.js]: Thanks! It's fixed.

**Comment by pmario** [comment]
@Jermolene Here is a version created with the latest pre-release code [katexdemo.zip](https://github.com/Jermolene/TiddlyWiki5/files/6938061/katexdemo.zip)

It seems to work as proposed. @8d1h You may create a post in the google group, so others users may test this. So we can see if there any problems

File: plugins/tiddlywiki/katex/config.tid [hunk 2]
```diff
@@ -0,0 +1,13 @@
+title: $:/plugins/tiddlywiki/katex/config
+
+<$list filter="[all[shadows+tiddlers]tag[$:/tags/katex/config]!has[draft.of]]">
```
Comment (Jermolene) [inline plugins/tiddlywiki/katex/config.tid:3]: For consistency, the tag should be `$:/tags/KaTeX/Config`

File: plugins/tiddlywiki/katex/wrapper.js [hunk 3]
```diff
@@ -16,6 +16,23 @@ var katex = require("$:/plugins/tiddlywiki/katex/katex.min.js"),
...
+katex.macros = {};
+katex.updateMacros = function() {
+	var tiddlers = $tw.wiki.getTiddlersWithTag("$:/tags/katex/macro"),
+		regex = /#\d/g, // Remove the arguments like #1#2
+		tid, macro, cmd;
+	for (var i=0; i < tiddlers.length; i++) {
+		tid = $tw.wiki.getTiddler(tiddlers[i]);
+		try {
+			macro = tid.fields["macro"];
+			macro = macro.replace(regex, "");
```
Comment (Jermolene) [inline plugins/tiddlywiki/katex/wrapper.js:28]: Why do we remove the `#1`, `#2` from the "macro" string?

**Review by Jermolene — COMMENTED** [review]
Thank you @8d1h, it's fantastic when we get such complete submissions from new contributors.

I have a couple of minor questions below. More significantly, I'm wondering if we might not be better off taking advantage of our existing macro/variable system.

The idea would be to define a KaTeX macro in the same way as an ordinary wikitext macro:

```
\define \ZZ() \mathbb{Z}
```

Such macros could be defined locally or globally.

The KaTeX wrapper would then construct the macros from the in-scope variables whose name starts with a backslash.

This approach would mean that macros could be dynamically redefined without needing to reload.

**Comment by 8d1h** [comment]
> The KaTeX wrapper would then construct the macros from the in-scope variables whose name starts with a backslash.
> 
> This approach would mean that macros could be dynamically redefined without needing to reload.

Thanks! I think that is indeed a better choice in theory. However, after a bit of thought, this would basically mean that we need to write a parser for the LaTeX syntax itself.

For example, the second macro I listed could be used as `\dd{\sin{x}}{x}` for dsin(x)/dx. We will then need to parse the two parts `\sin{x}` and `x` to be able to generate the correct result `\frac{d\sin{x}}{d x}`. The macro calls could also be nested, like d(dsinx/dx)/dx, then things could get a lot more complicated. So right now I'm just letting KaTeX do all the heavy-lifting :)

In short, this editor is just a nicer interface for defining macros in KaTeX. It would require a lot more work to properly parse the LaTeX syntax, and probably not worth it since we are already using KaTeX to do this for us...

File: plugins/tiddlywiki/katex/config.tid (hunk 2 above)
Comment (8d1h) [reply plugins/tiddlywiki/katex/config.tid:3]: Ok I will fix this.

File: plugins/tiddlywiki/katex/wrapper.js (hunk 3 above)
Comment (8d1h) [reply plugins/tiddlywiki/katex/wrapper.js:28]: The KaTeX macro variable does not accept inputs like `\dd#1#2`, however I think it is more natural to allow this in the definition, since usually one writes `\def\dd#1#2{\frac...}`. So now both `\dd` and `\dd#1#2` can be put in the macro editor, and KaTeX will only get `\dd` in either case.

**Comment by 8d1h** [comment]
Maybe I can put the macro editor in the KaTeX editor toolbar, besides the snippets, then we won't need to make definitions in individual files and can just add them from the toolbar.

**Comment by kookma** [comment]
@8d1h - each macro uses a new tiddler like 


`$:/plugins/tiddlywiki/katex/macros/macro`
`$:/plugins/tiddlywiki/katex/macros/macro 1`
`$:/plugins/tiddlywiki/katex/macros/macro 2`

I think in TW 5.0.2 you have options to create more meaningful name
also each tiddler has a new `macro` field. I may recommend to use caption which is a standard field here

I believe these tiddlers can be more semantic and tidy!

One crude proposal:
1. every macro has a name so you can use `$:/plugins/tiddlywiki/katex/macros/`+name
2. the KateX macro then can create from `$:/plugins/tiddlywiki/katex/macros/name` and what exists in text field!

**Comment by 8d1h** [comment]
> @8d1h - each macro uses a new tiddler like
> 
> `$:/plugins/tiddlywiki/katex/macros/macro`
> `$:/plugins/tiddlywiki/katex/macros/macro 1`
> `$:/plugins/tiddlywiki/katex/macros/macro 2`
> 
> I think in TW 5.0.2 you have options to create more meaningful name
> also each tiddler has a new `macro` field. I may recommend to use caption which is a standard field here
> 
> I believe these tiddlers can be more semantic and tidy!
> 
> One crude proposal:
> 
>     1. every macro has a name so you can use `$:/plugins/tiddlywiki/katex/macros/`+name
> 
>     2. the KaTeX macro then can be created from `$:/plugins/tiddlywiki/katex/macros/name` and what exists in text field!

Thanks for the comments! Yes you are right I was being lazy and trying to avoid naming them properly! :) They are now named properly after the macro, and I also changed the field from `macro` to `caption`.

**Comment by pmario** [comment]
@8d1h ... If you intend to add functions and / or tweak the project, you should make the PR a "draft". 

grafik

Once you are finished and the PR should be merged you convert it back to a PR that should be reviewed. This button is near the end of the conversation 

grafik

**Comment by Jermolene** [comment]
Thanks for your patience @8d1h

**Comment by kookma** [comment]
@8d1h - with the release of official KaTeX 0.13.18 I would recommend to add documentation for custom macros to katex official plugin page!

Also see https://talk.tiddlywiki.org/t/tiddlywiki-5-2-and-katex-0-13-18/704/7

**Comment by pmario** [comment]
> Also see https://talk.tiddlywiki.org/t/tiddlywiki-5-2-and-katex-0-13-18/704/7

I do like the "talk" links way more than GG links. They actually mean something.

**Comment by 8d1h** [comment]
> @8d1h - with the release of official KaTeX 0.13.18 I would recommend to add documentation for custom macros to katex official plugin page!

@kookma I just saw your TiddlyWiki of KaTeX demos, which is very nice! (I guess it's also useful for LaTeX users in general.) You should go ahead and update the official demo.

> Why not just `$:/plugins/tiddlywiki/katex/macros/ergun`? You can trim `\` from macro name and then add as suffix to `$:/plugins/tiddlywiki/katex/macros/`

Actually I was also thinking of doing that, but KaTeX actually supports redefining a single letter: e.g. you can create a macro that maps `a` (without backslash) to `bbbbb` (this is certainly a very bad practice), so I kept the backslash as part of the macro.

**Comment by 8d1h** [comment]
BTW in your macro example, I'd say for very complicated formula it's better to use the stamp functionality (the katex logo that can be seen from the editor), since one can insert the formula easily while still be able to edit it in place. The macros are most useful for small combinations of symbols that one uses a lot and might be tedious to type each time. For example I think this [wiki](https://mathematik.tiddlyspot.com/#%24%3A%2F_macros%2FKaTeX) has a very typical usecase

**Comment by kookma** [comment]
@8d1h - when click `add` or `import` it is better to clear the input box specially if the addition is successful!
Right now when you work with a long list it is very confusing and you had to see all macros again to see if your entry is added!

**Comment by 8d1h** [comment]
Yes you are right. Maybe one better solution is to add the option to sort the macros by creation time? So the last added ones will be at the end of the list.

**Comment by kookma** [comment]
> Yes you are right. Maybe one better solution is to add the option to sort the macros by creation time? So the last added ones will be at the end of the list.

Good! you can also add a button on header to sort them! by the way I may recommend to clear the box (here you have used a temp tiddler) after successful addition!
